File: 77.js
function maxSlidingWindow(nums, k) {
  const dq = []; // store indexes
  const res = [];

  for (let i = 0; i < nums.length; i++) {
    if (dq.length && dq[0] <= i - k) dq.shift();

    while (dq.length && nums[dq[dq.length - 1]] < nums[i]) {
      dq.pop();
    }

    dq.push(i);

    if (i >= k - 1) res.push(nums[dq[0]]);
  }

  return res;
}

// Driver
console.log(maxSlidingWindow([1,3,-1,-3,5,3,6,7], 3));

File: 77.py
from collections import deque

def maxSlidingWindow(nums, k):
    dq = deque()  # stores indexes
    result = []

    for i in range(len(nums)):
        # remove elements out of window
        if dq and dq[0] <= i - k:
            dq.popleft()

        # maintain monotonic decreasing queue
        while dq and nums[dq[-1]] < nums[i]:
            dq.pop()

        dq.append(i)

        # window forms
        if i >= k - 1:
            result.append(nums[dq[0]])

    return result


# Driver
nums = [1,3,-1,-3,5,3,6,7]
k = 3
print(maxSlidingWindow(nums, k))
